src/freertos_drivers/common/WiFiInterface.hxx

```diff
@@ -0,0 +1,273 @@
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @file WiFiInterface.hxx
+ *
+ * Common interface for WiFi operations.
+ *
+ * @author Stuart Baker
+ * @date 23 Jun 2025
+ */
+
+#ifndef _FREERTOS_DRIVERS_COMMON_WIFIINTERFACE_HXX_
+#define _FREERTOS_DRIVERS_COMMON_WIFIINTERFACE_HXX_
+
+#include <functional>
+#include <string>
+#include <vector>
+
+#include "utils/Singleton.hxx"
+#include "freertos_drivers/common/WifiDefs.hxx"
+
+/// Abstract interface for WiFi operations
+class WiFiInterface : public Singleton<WiFiInterface>
+{
+public:
+    /// Interface index by type.
+    enum Interface : uint8_t
+    {
+        IFACE_STA, ///< STA mode interface
+        IFACE_AP, ///< AP mode interface
+    };
+
+    /// Security types.
+    enum SecurityType : uint8_t
+    {
+        SEC_OPEN = 0, ///< open (no security)
+        SEC_WEP, ///< WEP security mode
+        SEC_WPA2, ///< WPA2 security mode
+    };
+
+    /// Network info, typically used in an access point scan.
+    struct NetworkEntry
+    {
+        /// Constructor.
+        /// @param ssid SSID of the access point
+        /// @param sec_type security type of the access point
+        /// @param rssi receive signal strength of the access point
+        NetworkEntry(const char *ssid, SecurityType sec_type, int rssi)
+            : ssid(ssid)
+            , secType(sec_type)
+            , rssi(rssi)
+        {
+        }
+
+        std::string ssid; ///< SSID of the access point
+        SecurityType secType; ///< security type of the access point
+        int rssi; ///< receive signal strength of the access point
+    };
+
+    /// Initialize the WiFi.
+    virtual void init()
+    {
+    }
+
+    /// Start the WiFi.
+    /// @param role device role
+    virtual void start(WlanRole role = WlanRole::DEFAULT_ROLE) = 0;
+
+    /// Stop the WiFi.
+    virtual void stop() = 0;
+
+    /// Get the started state of the WiFi.
+    /// @return true if started, else false
+    virtual bool is_started()
+    {
+        return started_;
+    }
+
+    /// Connect to access point.
```

Comment:
From @balazsracz: Please document whether this is a blocking call or an asynchronous call.
On the cc32xx this call seems to be blocking, i.e., it could take several seconds then return an error.
It's unclear to me on the esp32, i think the implementation in this PR is asynchronous (i.e., non-blocking) and the return value seems to be meaningless.

Reply:
I've addressed this fully through documentation and addition of a connect/disconnect user callback. @balazsracz Please have another look.

Reply:
looks good